Modification to region geom for performance

registerrier opened the pull request:
<!-- These are hidden commments. You can leave them or delete them. -->

<!-- Don't be shy! Github issues or pull requests are welcome any time. -->
<!-- But if you first want to chat, use the Gammapy Slack. -->
<!-- See https://gammapy.org/contact.html -->

**Description**
<!-- What is the motivation for this pull request? -->
<!-- Briefly: what changes are done here? -->
<!-- If this is to address a Github issue, metion it's number to create a link -->
<!-- Example: This PR is to fix issue #42 -->

This pull request corrects a bug in `SkyModel.contributes` where the `region_mask` was returning `False` inside the region leading the method to always return `False`. Consequently model spatial model evaluation was performed at each `npred` call.

Another change is made to `RegionGeom.center_skydir` to use `region.center` when it exists and the region is not compound. This also improves the run time. 

**Dear reviewer**
<!-- Let the reviewer and Gammapy team know what you want: -->
<!-- * Is this ready for review? Or is it work in progress and you want some feedback? -->
<!-- * Do you want ot go through review here? Or if someone just finish this up and merge it in? -->
<!-- Do you have any specific questions, e.g. about API or implementation? -->
<!-- Do you include a test executing new code you're adding (to make sure it runs)? -->
<!-- Do you include some documentation? Is it needed? -->

adonath reviewed (commented):
Thanks @registerrier! As `RegionGeom.region` is read-only and cannot be modified it's probably simpler and equally safe to just make `self._rectangle_bbox` or even `center_skydir` a `lazyproperty`?

adonath commented on gammapy/modeling/models/cube.py:
```diff
@@ -252,7 +252,7 @@ def contributes(self, mask, margin="0 deg"):
         if mask.geom.is_region and mask.geom.region is not None:
             geom = mask.geom.to_wcs_geom()
-            mask = geom.region_mask([mask.geom.region])
+            mask = ~geom.region_mask([mask.geom.region])
```
There is a related test-fail: https://github.com/gammapy/gammapy/pull/3484/checks?check_run_id=3414904505#step:9:1975

Maybe it's good to add a unit test for this as well...

registerrier commented:
I have added a unit test for the `SkyModel.contributes` with region mask input. 
I have also reverted to using the rectangle box center setting these as lazyproperties. 
Since this is very inefficient on a single call, one should check whether this is a bottleneck in some cases.

registerrier commented:
The problem here is actually different. In the `spectrum_1d` benchmark, the dataset are written and read from disk. There is a big in `regions` v0.5 which causes regions to become  `excluded` after `Regions.parse` in fits format. The `RegionGeom` after serialization is therefore an exclusion region. `model.contributes` cannot return `True` in this case, leading to the numerous evaluations causing the performance issue.
See: https://github.com/astropy/regions/issues/400

I will update the PR accordingly.
